### ISSUE TITLE
Ripser output intervals to file

### ISSUE THREAD

**Issue by tmelorc** [Ripser/ripser#10]
Is it possible to save all intervals computed by Ripser to an external file?

I am trying to edit the code to do this, but some intervals are not inserted in the file.

I was using `fopen()` and `fprintf` (eg. `fprintf(ripser_out, "%lu %.8f %.8f\n",dim,diameter,death);`) but for some random data set the output on screen is not the same as the one created in output file.

I have been working on this for more than 1 week but no progress.

It would be good to have an output file to be used within other programs. So a good output could have the format: `dim birth death` for intervals as `[a,b)` and `dim birth -1` for `[a, )`.

I have  no idea why my code does not work. Maybe because some limitation with `fopen`.

I appreciate any help.
Thanks in advance.

**Comment by sauln** [comment]
If you're willing to step outside of c++, I created a short script to convert the output into JSON using Python. 

You can find the gist here 
https://gist.github.com/sauln/da65fa1d699c9857bb9ba845973e6745

**Comment by tmelorc** [comment]
Thanks. I'd like to adapt the `cpp` code. It is easy to use some `grep` or `sed` to manipulate the original output and produce those format to export on a file, using `bash` for example. But I'd like to produce that from Ripser so I can only run it and get the output to use in other programs.
Anyway, thanks.

**Comment by sauln** [comment]
I get that. If you get it to work, I would appreciate the addition also.

**Comment by tmelorc** [comment]
After aksing help on StackOverflow, I got a solution and now I extended the `ripser.cpp` to `ripser_otf.cpp` (otf=OutputToFile). So, how can I propose the new version?

**Comment by sauln** [comment]
You want to submit your changes to this repo?  

You need to fork this repo, clone it to your computer, make the changes onto that copy of the repo, ad then push back up to your fork. Once that is done, github will ask if you want to submit a pull request.

Check out https://help.github.com/articles/fork-a-repo/

**Comment by tmelorc** [comment]
OK, thanks. I did a fork but I never used Git so I'll try to create a pull request, as necessary. That compare part is strange. Lets see.
Regards.

**Comment by sauln** [comment]
Git can definitely be confusing, even after using it for a long time. 

We’re you able to commit your new changes and push the commits up to github?

**Comment by tmelorc** [comment]
Too complicated. I have no idea how to use this. 
If you wish, please, try to open from here:
[https://github.com/tmelorc/ripser](url)

Please, feel free to help to pull.

**Comment by sauln** [comment]
That pull request looks good. I added the ability to input a filename as an argument.  

I made a pull request to your branch. If you accept that request, then (I'm pretty sure) it will automatically update your original pull request to incorporate those changes.

**Comment by tmelorc** [comment]
Well, I did. I'm not sure if it right, but let's see. I'll try to get your version to my computer, using git again. Hopefully it will work. In case of problem, I have backup.

**Comment by sauln** [comment]
Looks like it worked correctly.  Now that you've merged the pull request, all you should have to do is
```
git pull origin master
```

**Comment by tmelorc** [comment]
If I do this, the new version from git would be downloaded to my computer?

**Comment by sauln** [comment]
Yes.  How did you originally get the code down to your computer? You should have a local copy of the repo, and running `git pull` will pull the new updates from `origin master` where `origin` is your fork in the cloud and `master` is the branch or your fork.

**Comment by tmelorc** [comment]
Usually I download the file directly from the git, using the raw file. I'll try `git pull`.

**Comment by tmelorc** [comment]
OK, now it works. I got the new version. I've just compiled it and run. It works. But the output is not default. The option --output you created is necessary to have the output on file. 
What do you think to make it default? Then if we pass --output option we can choose the output file name.

Thanks for your help.

**Comment by sauln** [comment]
as it is now, if you don't pass a file name, then it won't print out a file. I think that'd be better for a default use case, rather than always create a new file. You might want something different yourself though.

**Comment by tmelorc** [comment]
Good. After two weeks trying to print that, finally it woks. 
Regards.

**Comment by sauln** [comment]
It looks like every time you are commiting, it updates the pull request. You could close the request, create a new branch, and then issue the pull request from that branch. 

That way you can keep pushing new commits without it effecting the essential pull request.

**Comment by tmelorc** [comment]
Well, sorry. This git is really very complicated.
I closed the pull request. I have no idea how to use new branch. So I think I'll stop committing for a while, since I was only putting some examples more. Nothing important.
Thanks for helping.
Regards.

**Comment by sauln** [comment]
No worries. It is pretty complicated and there is a ton to learn. It gets pretty complicated fast. 

You can create a branch using ‘git checkout -b BRANCHNAME’ and then commiting changes. You would then push to github using ‘git push origin BRANCHNAME’.

**Comment by tmelorc** [comment]
What should be a good BRANCHNAME? Some name with my user name, or with Ripser + my_user name? Or ripser_tests, or...? This name will be used only by me or should be relevant to others?

**Comment by sauln** [comment]
Usually you want it to capture the intent of your work.  Something like “output-examples” if that’s what you’re working on. No need for your name since all the commits will have that already also. Same with the library name, it would be redundant.

**Comment by toggled** [comment]
Hi all, 
Is this issue still open? Since I see the @tmelorc 's pull request still open.

**Comment by sauln** [comment]
The feature is sitting on @tmelorc's fork and in PR #11. So the issue is resolved, but not merged into master.

[also, if you haven't seen it, checkout @ctralie's fork, we have some nice Python bindings over there ]

**Comment by tmelorc** [comment]
Since the licence was changed some time ago, it would be good to have this feature in the original program (I'm not sure if forked material is allowed anymore).

**Comment by tmelorc** [comment]
I'm closing this issue since now I have some better knowledge how Github works and maybe I can deal with a better way to propose some changes. Thanks.
